realm/realm-library/src/main/java/io/realm/internal/Context.java

```diff
@@ -16,32 +16,60 @@
 
 package io.realm.internal;
 
-import java.lang.ref.Reference;
 import java.lang.ref.ReferenceQueue;
 import java.util.ArrayList;
-import java.util.HashMap;
 import java.util.List;
 
 public class Context {
 
+    // Pool to hold the phantom references.
+    // The size of array for storing phantom references will never decrease. Instead, we use another array to hold the
+    // index of the free slot. When adding the reference, pick the last index from freeIndexList and put the reference
+    // to the corresponding slot.
+    private static class ReferencesPool {
+        ArrayList<NativeObjectReference> arrayList = new ArrayList<>();
+        ArrayList<Integer> freeIndexList = new ArrayList<Integer>();
+
+        void add(NativeObjectReference ref) {
+            if (arrayList.size() <= ref.refIndex) {
+                arrayList.add(ref);
+            } else {
+                arrayList.set(ref.refIndex, ref);
+            }
+        }
+
+        Integer getFreeIndex() {
+            Integer index;
+            if (freeIndexList.isEmpty()) {
+                index = arrayList.size();
+            } else {
+                int freeIndexToRemove = freeIndexList.size() - 1;
+                index = freeIndexList.get(freeIndexToRemove);
+                freeIndexList.remove(freeIndexToRemove);
```

Comment:
these three lines can be

```
index = freeIndexList.remove(freeIndexList.size() - 1);
```

and if we use `ArrayDeque` instead of `ArrayList`, we can write it as

```
index = freeIndexList.removeLast()
```